Refactor difficulty icons to not suck

peppy opened the pull request:
Something I've been putting off for a while. The hierarchy of these icons and the very sparse and different usages made it a nightmare to deal with.

- `DifficultyIcon` is now a very simple implementation with a bindable to update the rating externally, and some sane defaults.
- `CalculatingDifficultyIcon` is responsible for doing automatic calculation when on-screen.
- `DifficultyRetriever` is a stand-alone component that handles the retrieval. I figure this may be useful in other contexts in the future. Or not. At least it's well defined and documented now.
- I was able to delete/combine some classes along the way (`GroupedFilterable` and `Filterable` combined, and some weird `MatchModeType` nuked).

peppy committed:
Move `DifficultyRetrieve` to own class and split both pathways into separate constructors

peppy committed:
Tidy up `DifficultyIcon`

peppy committed:
Remove unused `ModeTypeInfo` class

peppy committed:
Split out simple `DifficultyIcon` with no calculation overhead and update usages

peppy committed:
Tidy up all remaining usages

peppy commented on osu.Game/Beatmaps/Drawables/DifficultyRetriever.cs:
```diff
@@ -0,0 +1,83 @@
+            this.beatmapInfo = beatmapInfo;
+        }
+
+        /// <summary>
+        /// Construct a difficulty retriever that is calculated only once for the specified ruleset / mod combination.
+        /// This will not track global ruleset and mod changes.
+        /// </summary>
+        /// <param name="beatmapInfo">The beatmap to use for calculation.</param>
+        /// <param name="ruleset">The ruleset to use for calculation.</param>
+        /// <param name="mods">The mods to use for calculation.</param>
+        public DifficultyRetriever(IBeatmapInfo beatmapInfo, IRulesetInfo ruleset, IReadOnlyList<Mod> mods)
+        {
+            this.beatmapInfo = beatmapInfo;
+            this.ruleset = ruleset;
+            this.mods = mods;
+        }
```
I'm not sure if this is worth keeping around. It is not used anywhere right now. It is also the only usage of:

https://github.com/ppy/osu/blob/f8830c6850128456266c82de83273204f8b74ac0/osu.Game/Beatmaps/BeatmapDifficultyCache.cs#L104-L117

frenzibyte replied:
I think it'll be useful for cases like queued up `DrawableRoomPlaylistItem` which are not currently selected but need to display the correct SR icon in the item.

peppy replied:
Is now gone anyway.

frenzibyte commented on osu.Game/Overlays/BeatmapSet/BeatmapPicker.cs:
```diff
@@ -274,8 +274,9 @@ public DifficultySelectorButton(APIBeatmap beatmapInfo)
                         }
                     },
-                    icon = new DifficultyIcon(beatmapInfo, shouldShowTooltip: false)
+                    icon = new DifficultyIcon(beatmapInfo.Ruleset)
                     {
+                        Current = { Value = new StarDifficulty(beatmapInfo.StarRating, 0) },
```
Is it intended that this doesn't just use `DifficultyIcon(BeatmapInfo)` and disable tooltip explicitly?

peppy replied:
This was before I added the `ShowTooltip`. I'll update accordingly.

frenzibyte commented on osu.Game/Beatmaps/Drawables/DifficultyIcon.cs:
```diff
@@ -124,69 +111,23 @@ private void load(OsuColour colours)
 
-        ITooltip<DifficultyIconTooltipContent> IHasCustomTooltip<DifficultyIconTooltipContent>.GetCustomTooltip() => new DifficultyIconTooltip();
+        ITooltip<DifficultyIconTooltipContent> IHasCustomTooltip<DifficultyIconTooltipContent>.
+            GetCustomTooltip() => new DifficultyIconTooltip();
```
Was this automated by rider? I've also seen this happening locally on any file I can touch recently, not sure if something changed in the configuration somewhere.

peppy replied:
I'd say so.

frenzibyte reviewed (commented):
~~Upon testing, it seems difficulties that are beyond the song select screen area but not completely offscreen will remain unpopulated, making it look off:~~

https://user-images.githubusercontent.com/22781491/175451666-01086c42-2a0d-4139-a6ad-d50ac31cb735.mp4

~~(notice the difficulty cards at the bottom right of the screen updating SR the moment they enter song select range)~~

---

Never mind, I thought this PR introduced the usage of `DelayedLoadUnloadWrapper` for processing difficulty, while it's actually already present in master.

peppy committed:
Remove unused parameters

peppy committed:
Use new property to specify no tooltip, rather than passing only ruleset

peppy committed:
Update retriever to be relatively sized

peppy committed:
Nuke calculating everything

The whole component is pointless so I'm just going to nuke for now I
guess. Kind of makes the whole refactor effort pointless but oh well?

To expand on this, the implementation was actually incorrect as pointed
out at https://github.com/ppy/osu/pull/18819#pullrequestreview-1017886035.

peppy commented:
I've fixed all the reported issues. See https://github.com/ppy/osu/pull/18819/commits/28837693e5a72ccf692df76a7bc231a473987c4b commit message for important dialogue.

peppy committed:
Merge branch 'master' into difficulty-icon-refactor-pass
